core/trino-main/src/main/java/io/trino/operator/join/JoinHashSupplier.java

```diff
@@ -33,6 +34,16 @@
 public class JoinHashSupplier
         implements LookupSourceSupplier
 {
+    /**
+     * The `BigintPagesHash` performance comes with a memory consumption cost,
+     * having a dedicated long[] array to store values. If the number of positions
+     * gets too high, the memory consumption starts to grow significantly, so
+     * we use default implementation if it exceeds this value.
+     * This value is purposefully identical to that of IncrementalLoadFactorHashArraySizeSupplier#THRESHOLD_50,
+     * as higher load factor means more excessive memory consumption
+     */
+    private static final int BIGINT_SINGLE_CHANNEL_MAX_ADDRESSES = 1 << 20; // 1024576
```

Comment:
So in theory the memory utilization shouldn't increase by more than 8MB, right?

I wonder how difficult would it be to avoid storing values twice (once in a block and once in an array)?

Reply:
Actually a bit less. We stored a single byte hash per hash bucket before this PR. Now we store 8 bytes per actual position. The load factor for < 1M positions is 0.5 meaning an average load of 0.375. So we swapped `x` bytes for `8*x*0.375`=`3*x` bytes.
With number f positions > 1M, the load factor is 0.75, which translates to `4.5*x` bytes.
The previous version that failed the verifier added a constant 8 bytes, regardless of the load factor

Reply:
> I wonder how difficult would it be to avoid storing values twice (once in a block and once in an array)?

This is, unfortunately, really tricky, since the value blocks are used by other things like filters and sorting and the addressing convention is consistent across all channels, not only joined ones.